n3fit/src/n3fit/layers/mask.py

```diff
@@ -25,23 +25,27 @@ class Mask(MetaLayer):
     def __init__(self, bool_mask=None, c=None, axis=None, **kwargs):
         if bool_mask is None:
             self.mask = None
+            self.last_dim = -1
         else:
             self.mask = op.numpy_to_tensor(bool_mask, dtype=bool)
+            self.last_dim = count_nonzero(bool_mask[0,...])
         self.c = c
         self.axis = axis
         super().__init__(**kwargs)
 
     def build(self, input_shape):
         if self.c is not None:
             initializer = MetaLayer.init_constant(value=self.c)
+            output_shape = list(input_shape)
+            output_shape[-1] = self.last_dim
             self.kernel = self.builder_helper(
-                "mask", (1,), initializer, trainable=False
+                "mask", output_shape, initializer, trainable=False
```

Comment:
The implicit shape argument leads to a tensorflow graph execution error
`Node: 'meta_model_1/tf.concat_12/concat'
ConcatOp : Dimension 1 in both shapes must be equal: shape[0] = [1,3,3] vs. shape[7] = [1,1,0]
	 [[{{node meta_model_1/tf.concat_12/concat}}]] [Op:__inference_losses_fun_110454]`
(this is for the NNPDF4.0 runcard with 3 parallel replicas).

Can we expect tensorflow to be able to infer the output shape from the mask? I would be surprised if it can, I haven't been able to find useful documentation about this, so I tried to be as explicit as I could be.

Reply:
Ok, apparently tensorflow inferred the output shape from the multiplication constant (which was no longer a scalar but a const tensor, and the multiplication was being done point-wise). I removed that and made the output shape explicit upon reshaping in commit 303024e7a1cade4150caa2154a36f78d19d04257